cmd/kubeadm/app/util/staticpod/utils_linux.go

```diff
@@ -0,0 +1,158 @@
+
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package staticpod
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	certphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/certs"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util/users"
+	"k8s.io/utils/pointer"
+)
+
+type pathOwerAndPermissionsUpdaterFunc func(path string, uid, gid int64, perms uint32) error
+
+// RunComponentAsNonRoot updates the pod manifest and the hostVolume permissions to run as non root.
+func RunComponentAsNonRoot(componentName string, pod *v1.Pod, usersAndGroups *users.UsersAndGroups, cfg *kubeadmapi.ClusterConfiguration) error {
+	switch componentName {
+	case kubeadmconstants.KubeAPIServer:
+		return runKubeAPIServerAsNonRoot(
+			pod,
+			usersAndGroups.Users.ID(kubeadmconstants.KubeAPIServerUserName),
+			usersAndGroups.Groups.ID(kubeadmconstants.KubeAPIServerUserName),
+			usersAndGroups.Groups.ID(kubeadmconstants.ServiceAccountKeyReadersGroupName),
+			users.UpdatePathOwnerAndPermissions,
+			cfg,
+		)
+	case kubeadmconstants.KubeControllerManager:
+		return runKubeControllerManagerAsNonRoot(
+			pod,
+			usersAndGroups.Users.ID(kubeadmconstants.KubeControllerManagerUserName),
+			usersAndGroups.Groups.ID(kubeadmconstants.KubeControllerManagerUserName),
+			usersAndGroups.Groups.ID(kubeadmconstants.ServiceAccountKeyReadersGroupName),
+			users.UpdatePathOwnerAndPermissions,
+			cfg,
+		)
+	case kubeadmconstants.KubeScheduler:
+		return runKubeSchedulerAsNonRoot(
+			pod,
+			usersAndGroups.Users.ID(kubeadmconstants.KubeControllerManagerUserName),
+			usersAndGroups.Groups.ID(kubeadmconstants.KubeControllerManagerUserName),
+			users.UpdatePathOwnerAndPermissions,
+		)
+	}
+	return errors.New(fmt.Sprintf("component name %q is not valid", componentName))
+}
+
+// runKubeAPIServerAsNonRoot updates the pod manifest and the hostVolume permissions to run kube-apiserver as non root.
+func runKubeAPIServerAsNonRoot(pod *v1.Pod, runAsUser, runAsGroup, supplementalGroup *int64, updatePathOwnerAndPermissions pathOwerAndPermissionsUpdaterFunc, cfg *kubeadmapi.ClusterConfiguration) error {
+	saPublicKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPublicKeyName)
+	if err := updatePathOwnerAndPermissions(saPublicKeyFile, *runAsUser, *runAsGroup, 0600); err != nil {
```

Comment:
non-blocking: should we add local constants for 0600 and 0640?
e.g. `perm0600`

Reply:
I am seeing the current pattern used in k/k a lot, I don't feel too strongly about it so your call.

https://github.com/kubernetes/kubernetes/search?p=2&q=0600

Reply:
ok, let's leave it the way it is.